infer/apply_defaults.go

```diff
@@ -0,0 +1,354 @@
+			return false, err
+		}
+		if didSetRec {
+			didSet = true
+		}
+	}
+	return didSet, nil
+}
+
+// walk is responsible for calling applyDefaults on all structs in reachable from value.
+func (d *defaultsWalker) walk(v reflect.Value) (didSet bool, _ error) {
+	tDeferenced := v.Type()
+	for tDeferenced.Kind() == reflect.Pointer {
+		tDeferenced = tDeferenced.Elem()
+	}
+	switch tDeferenced.Kind() {
+	// We apply defaults to each value in slice.
+	case reflect.Slice:
+		// Dereference to the underlying slice
+		for v.Kind() == reflect.Pointer && !v.IsNil() {
+			v = v.Elem()
+		}
+		// Either we have a type *[]T, **[]T, etc. and a pointer is nil,
+		// or we have reached the slice and the slice itself is nil. Both
+		// cases prevent us finding any more structs, so we are done.
+		if v.IsNil() {
+			return didSet, nil
+		}
+		for i := 0; i < v.Len(); i++ {
+			didSetRec, err := d.walk(v.Index(i))
+			if err != nil {
+				return false, err
+			}
+			if didSetRec {
+				didSet = true
+			}
+		}
+		return didSet, nil
+
+	case reflect.Array:
+		// Dereference to the underlying slice
+		for v.Kind() == reflect.Pointer {
+			if v.IsNil() {
+				return false, nil
+			}
+			v = v.Elem()
+		}
+
+		// Arrays cannot be nil, so we don't (and can't) perform a nil
+		// check here.
+		for i := 0; i < v.Len(); i++ {
+			didSetRec, err := d.walk(v.Index(i))
+			if err != nil {
+				return false, err
+			}
+			if didSetRec {
+				didSet = true
+			}
+		}
+		return didSet, nil
+
+	case reflect.Map:
+		// Dereference to the underlying map
+		for v.Kind() == reflect.Pointer && !v.IsNil() {
+			v = v.Elem()
+		}
+		// Either we have a type *map[K]V, **map[K]V, etc. and a pointer
+		// is nil, or we have reached the map and the map itself is
+		// nil. Both cases prevent us finding any more structs, so we are
+		// done.
+		if v.IsNil() {
+			return false, nil
+		}
+		iter := v.MapRange()
+		for iter.Next() {
+			value := reflect.New(iter.Value().Type()).Elem()
+			value.Set(iter.Value())
+			s := hydratedValue(value)
+			didSetRec, err := d.walk(s)
+			if err != nil {
+				return false, err
+			}
+			if didSetRec {
+				v.SetMapIndex(iter.Key(), s)
+				didSet = true
+			}
+		}
+		return didSet, nil
+	case reflect.Struct:
+		// Copying what Go SDKs do, we only populate structs that are accessible
+		// by value or by filled pointers.
+		copy := reflect.New(v.Type()).Elem()
+		copy.Set(v)
+		var s reflect.Value
+		if d.hydrate(tDeferenced) {
+			// We should hydrate the value and apply defaults to it.
+			s = hydratedValue(copy)
+		} else {
+			// We should fill a live value, but not hydrate a new one.  Set s
+			// to the copy, assuming that we will have a non-nil struct.
+			s = copy
+
+			// Now check that we have a non-nil struct.
+			for copy.Kind() == reflect.Pointer {
+				if copy.IsNil() {
+					return false, nil
+				}
+				copy = copy.Elem()
+			}
+		}
+		didSet, err := d.apply(derefNonNil(s))
+		if err != nil {
+			return false, err
+		}
+		if didSet {
```

Comment:
didSet tracking is optimizing away something here?

Reply:
We explore the full space of our input's fields to find where defaults should be applied, but we don't want to set up structs that were `(*struct{...})(nil)` unless the struct has a default value that should be applied. The pseudocode looks like this:

```go
if typeOf(field) == *T && field == nil {
    var copy &T{}
    didSet := apply(copy)
    if didSet {
        field = copy
    }
    // Otherwise we leave field to be nil
}
```